### ISSUE TITLE
Previously working code doesn't work on nightly (zero-sized types)

### ISSUE THREAD

**Issue by Karrq** [rust-lang/rust#77563]
I tried this code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d0881f2d1ffe9e1a71cc8d6c815a6019)):

```rust
struct ZST;
impl ZST {
  pub const fn new() -> Self { Self }

  pub const fn static_ref() -> &'static Self {
    &Self::new()
  }
}
```

I expected to see this happen: no errors

Instead, this happened: compiler errored with:

```
error[E0515]: cannot return reference to temporary value
 --> src/lib.rs:6:5
  |
6 |     &Self::new()
  |     ^-----------
  |     ||
  |     |temporary value created here
  |     returns a reference to data owned by the current function
```

### Meta

`rustc --version --verbose`:
```
rustc 1.49.0-nightly (beb5ae474 2020-10-04)
binary: rustc
commit-hash: beb5ae474d2835962ebdf7416bd1c9ad864fe101
commit-date: 2020-10-04
host: x86_64-unknown-linux-gnu
release: 1.49.0-nightly
LLVM version: 11.0
```

-----

This doesn't error in stable or beta, just nightly.
I worked this around by just returning the ZST (`&Self`) instead of calling the constructor (`&Self::new()`).

**Comment by SNCPlay42** [comment]
This code errors for me on stable on the playground.

**Comment by Karrq** [comment]
Found out why, I attacked the wrong playground link, where I removed the `const` keywords.

[This one](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d0881f2d1ffe9e1a71cc8d6c815a6019) is the correct one. It compiles both in stable and beta, but not on nightly.

**Comment by SNCPlay42** [comment]
I believe this was an expected result of #75502 (and a `cargo bisect-rustc` points to its rollup, #76912).

As a workaround, this code works:

```rust
struct ZST;
impl ZST {
  pub const fn new() -> Self { Self }

  pub const fn static_ref() -> &'static Self {
    const VAL: ZST = ZST::new();
    &VAL
  }
}
```

**Comment by RalfJung** [comment]
Cc @rust-lang/wg-const-eval 

This code used to work accidentally, and stopped working when https://github.com/rust-lang/rust/issues/75586 got fixed.

**Comment by Karrq** [comment]
Alright, it's fine. This code was written way back in Sept 2019.

I'd like to point out that there exists a different workaround:
```
struct ZST;
impl ZST {
  pub const fn new() -> Self { Self }

  pub const fn static_ref() -> &'static Self {
    &Self
  }
}
```

as you can see there's no more function call.

The only thing is that moving forward this will break code that is on stable if I'm not mistaken, right?

**Comment by oli-obk** [comment]
> The only thing is that moving forward this will break code that is on stable if I'm not mistaken, right?

yes

**Comment by camelid** [comment]
Removing regression labels since this is an intended regression.

**Comment by camelid** [comment]
(I edited the issue to show the correct code and the full error.)

**Comment by oli-obk** [comment]
Closing as working-as-intended. Sorry about the accidental stabilization. We're working on more convenient ways to do the same thing.
